File: 47. Permutations II/47. Permutations II.dart
class Solution {
  List<List<int>> permuteUnique(List<int> nums) {
    List<List<int>> result = [];
    nums.sort();
    _backtrack(nums, 0, result);
    return result;
  }
  
  void _backtrack(List<int> nums, int start, List<List<int>> result) {
    if (start == nums.length) {
      result.add(List.from(nums));
      return;
    }
    
    Set<int> used = {};
    for (int i = start; i < nums.length; i++) {
      if (used.contains(nums[i]) || (i > start && nums[i] == nums[start])) continue;
      used.add(nums[i]);
      _swap(nums, start, i);
      _backtrack(nums, start + 1, result);
      _swap(nums, start, i);
    }
  }
  
  void _swap(List<int> nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
}

File: 47. Permutations II/47. Permutations II.rs
use std::collections::HashSet;

impl Solution {
    pub fn permute_unique(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result = Vec::new();
        let mut nums = nums;
        nums.sort_unstable();
        Self::backtrack(&mut nums, 0, &mut result);
        result
    }
    
    fn backtrack(nums: &mut Vec<i32>, start: usize, result: &mut Vec<Vec<i32>>) {
        if start == nums.len() {
            result.push(nums.clone());
            return;
        }
        
        let mut used = HashSet::new();
        for i in start..nums.len() {
            if used.contains(&nums[i]) || (i > start && nums[i] == nums[start]) {
                continue;
            }
            used.insert(nums[i]);
            nums.swap(start, i);
            Self::backtrack(nums, start + 1, result);
            nums.swap(start, i);
        }
    }
}